mmdet/core/hook/neptunelogger_hook.py

```diff
@@ -0,0 +1,305 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from __future__ import annotations
+import os.path as osp
+from typing import Dict, Optional
+
+from mmdet.core.visualization.image import imshow_det_bboxes
+
+try:
+    import neptune.new as neptune
+    from neptune.new.types import File
+except ImportError:
+    raise ImportError('Neptune client library not installed.'
+                      'Please refer to the installation guide:'
+                      'https://docs.neptune.ai/setup/installation/')
+
+import mmcv
+import mmcv.runner.hooks as mmvch
+import numpy as np
+from mmcv import Config
+from mmcv.runner import HOOKS, EpochBasedRunner, IterBasedRunner
+from mmcv.runner.dist_utils import master_only
+from mmcv.runner.hooks.checkpoint import CheckpointHook
+
+from mmdet import version
+from mmdet.core import DistEvalHook, EvalHook
+
+
+@HOOKS.register_module()
+class NeptuneHook(mmvch.logger.neptune.NeptuneLoggerHook):
+    """Log metadata to Neptune.
+
+    This hook will automatically log training or evaluation metadata
+    to Neptune.ai.
+    Args:
+        init_kwargs (dict): A dict passed to neptune.init_run to initialize
+        a Neptune run.
+            Please refer to https://docs.neptune.ai/api/neptune/#init_run
+            for possible key-value pairs.
+        api_token (str): Neptune API token.
+            If None, the environment variable NEPTUNE_API_KEY will be used.
+        project (str): Neptune project name.
+            If None, the environment variable NEPTUNE_PROJECT will be used.
+        interval (int): Checking interval (every k epochs/iterations).
+            Default: 10
+        base_namespace (str): Namespace under which all the logs will
+        be present.
+            Default: "training"
+        log_model (bool): If the Hook should save the final checkpoint.
+            Requires CheckpointHook.
+            Default: False
+        log_checkpoint (bool): If the Hook should log all the checkpoints
+        generated by CheckpointHook.
+            Requires CheckpointHook.
+            Default: False
+        num_eval_predictions (int): Maximum number of eval predictions to log.
+            Requires EvalHook.
+            Default: 50
+    """
+
+    INTEGRATION_VERSION_KEY = 'source_code/integrations/neptune-mmdetection'
+
+    def __init__(self,
+                 *,
+                 init_kwargs: Optional[Dict] = None,
+                 api_token: str = None,
+                 project: str = None,
+                 interval: int = 10,
+                 base_namespace: str = 'training',
+                 log_model: bool = False,
+                 log_checkpoint: bool = False,
+                 num_eval_predictions: int = 5,
+                 **kwargs) -> None:
+        super().__init__(interval=interval, **kwargs)
+
+        if init_kwargs:
+            api_token = init_kwargs.pop('api_token', None) or api_token
```

Comment:
Should we error in the case where `api_token` is passed through both `api_token` and `init_kwargs`?

Reply:
Maybe let's error if they are not the same? What do you think?

Reply:
I think this makes sense. In case of lightning we overwrite the `init_kwargs` if we have received `api_token` or `project` as an argument

https://github.com/Lightning-AI/lightning/blob/65eb3e7c456dc572d909362157ee2d29e99e2276/src/lightning/pytorch/loggers/neptune.py#L289-L296